drivers/dai/intel/ssp/ssp.c

```diff
@@ -0,0 +1,2021 @@
+
+	if (direction == DAI_DIR_CAPTURE) {
+		LOG_INF("%s RX", __func__);
+	} else {
+		LOG_INF("%s TX", __func__);
+	}
+
+	ssp->state[direction] = DAI_STATE_PAUSED;
+}
+
+static int dai_ssp_trigger(const struct device *dev, enum dai_dir dir,
+			   enum dai_trigger_cmd cmd)
+{
+	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
+	struct dai_intel_ssp_pdata *ssp = dai_get_drvdata(dp);
+	int array_index = SSP_ARRAY_INDEX(dir);
+
+	LOG_INF("%s cmd %d", __func__, cmd);
+
+	switch (cmd) {
+	case DAI_TRIGGER_START:
+		if (ssp->state[array_index] == DAI_STATE_PAUSED ||
+		    ssp->state[array_index] == DAI_STATE_PRE_RUNNING) {
+			dai_ssp_start(dp, array_index);
+		}
+		break;
+	case DAI_TRIGGER_STOP:
+		dai_ssp_stop(dp, array_index);
+		break;
+	case DAI_TRIGGER_PAUSE:
+		dai_ssp_pause(dp, array_index);
+		break;
+	case DAI_TRIGGER_PRE_START:
+		dai_ssp_early_start(dp, array_index);
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static const struct dai_config *dai_ssp_config_get(const struct device *dev, enum dai_dir dir)
+{
+	struct dai_config *params = (struct dai_config *)dev->config;
+	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
+	struct dai_intel_ssp_pdata *ssp = dai_get_drvdata(dp);
+
+	params->rate = ssp->params.fsync_rate;
+
+	if (dir == DAI_DIR_PLAYBACK) {
+		params->channels = popcount(ssp->params.tx_slots);
+	} else {
+		params->channels = popcount(ssp->params.rx_slots);
+	}
+
+	params->word_size = ssp->params.sample_valid_bits;
+
+	return params;
+}
+
+static int dai_ssp_config_set(const struct device *dev, const struct dai_config *cfg,
+			      const void *bespoke_cfg)
+{
+	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
+
+	if (cfg->type == DAI_INTEL_SSP) {
+		return dai_ssp_set_config_tplg(dp, cfg, bespoke_cfg);
+	} else {
+		return dai_ssp_set_config_blob(dp, bespoke_cfg);
+	}
+}
+
+static const struct dai_properties *dai_ssp_get_properties(const struct device *dev,
+							   enum dai_dir dir, int stream_id)
+{
+	struct dai_intel_ssp *dp = (struct dai_intel_ssp *)dev->data;
+	struct dai_intel_ssp_pdata *ssp = dai_get_drvdata(dp);
+	struct dai_properties *prop = &ssp->props;
+	int array_index = SSP_ARRAY_INDEX(dir);
+
+	prop->fifo_address = dp->plat_data.fifo[array_index].offset;
+	prop->dma_hs_id = dp->plat_data.fifo[array_index].handshake;
+
+	if (ssp->clk_active & SSP_CLK_BCLK_ACTIVE) {
+		prop->reg_init_delay = 0;
+	} else {
+		prop->reg_init_delay = ssp->params.bclk_delay;
+	}
+
+	LOG_INF("%s dai_index %u", __func__, dp->index);
+	LOG_INF("%s fifo %u", __func__, prop->fifo_address);
+	LOG_INF("%s handshake %u", __func__, prop->dma_hs_id);
+	LOG_INF("%s init delay %u", __func__, prop->reg_init_delay);
+
+	return prop;
+}
+
+static int dai_ssp_probe(struct dai_intel_ssp *dp)
+{
+	struct dai_intel_ssp_pdata *ssp;
+
+	if (dai_get_drvdata(dp)) {
+		return -EEXIST; /* already created */
+	}
+
+	/* allocate private data */
+	ssp = k_calloc(1, sizeof(*ssp));
```

Comment:
this is in fact interesting. I assume you have to increase `CONFIG_HEAP_MEM_POOL_SIZE` for this to work. Then - is the returned address cached or uncached? We use uncached so far. It seems to me, that this will be allocated from the `.noinit` section and it will be uncached too